### ISSUE TITLE
Standardize way to map fields on DDD elements

### ISSUE THREAD

**Issue by qejk** [meteor-space/event-sourcing#48]
Currently, there are two was to map(describe) fields in DDD elements, via: methods and objects:

`Entity`: require to map fields as a method
`ValueObject`: require to map fields as a method or object
`Aggregates/Process`: require objects

If we look on another DDD elements: `Commands`, `Events` - both store fields as objects.

**Comment by rhyslbw** [comment]
I have a feeling we need to use a method for objects that are extended from `Space.Struct` including `Entity` and `ValueObject`. @DominikGuzei is that right?

**Comment by rhyslbw** [comment]
Also need to consider if [this](https://github.com/meteor-space/messaging/blob/ff0dfa578b2fcd39f167f3c51afc6d9ac7011ae3/source/mixins/serializable.js#L107) backwards compatibility class should be removed now, freeing up the mixin to be named `Space.messaging.Serializable'

**Comment by DominikGuzei** [comment]
We just have to mixin `Space.messaging.SerializableMixin` into `Space.eventSourcing.Aggregate` and update the internal code so that it uses the `fields` method instead. The switch to remove the compatibilitty class is a bit bigger and needs careful consideration.

**Comment by rhyslbw** [comment]
Consideration for upgrade path, or internal? Right now it's a little disjointed as we have an extended object that's really suited to being a trait that's added via a mixin, so can't use the obvious name in the current API.

**Comment by DominikGuzei** [comment]
We just have to check where it is used and update all places :wink:

**Comment by rhyslbw** [comment]
If that's the only issue I'd suggest we proceed with that. I can do it this evening so we can release messaging without a delay. Now is the time for breaking things if there's a good reason :smile: 

@qejk Thoughts?

**Comment by DominikGuzei** [comment]
Haha, yeah it seems like it's never a good time to break things :wink: but yes, let's do that.

**Comment by qejk** [comment]
Yes! Like we spoke @DominikGuzei `Aggregates/Processes` also should be serializable `Space.messaging.SerializableMixin`. This will bring very niffty  concept later for interacting with app layer if will start adding `Repositories` to the mix. Long story short:

We store on MongoDB our `Aggregate` via Repository (not the one from ES, related to our domain part) and there use `toData()`. 

Then afterwards, when we need to talk with persistence (via `find/findOne` etc) and need whole/partial document - we recreate it with `toData()` and again we can interact with it normally like with EJSON's one (and still preserve sanity in doc structure yey). 

**IMPORTANT**: will it work for partial data that have mappings in fields as non `Match.Optional`(so required)?

**Comment by rhyslbw** [comment]
FYI the mixin is now named [Space.messaging.Ejsonable](https://github.com/meteor-space/messaging/blob/release/3.0.0/source/mixins/ejsonable.js#L20) since it's a hard implementation for EJSON.

> IMPORTANT: will it work for partial data that have mappings in fields as non Match.Optional(so required)?

If I'm understanding correctly, it's all based on your match statement per field.

**Comment by darko-mijic** [comment]
Are we sure we want to refactor `Space.messaging.Command` and `Space.messaging.Event` also?

**Comment by DominikGuzei** [comment]
It's not so much about refactoring the names of the classes but splitting up the domain related parts (`sourceId`, `timestamp`, `targetId` etc.) into a separate `Space.domain.Event` / `Command` class.

**Comment by rhyslbw** [comment]
Ok so I agree here that switching the `fields` prop on `Aggregate` to a method makes sense, but we should retain the ability in the style to use an object for fields in a `define` static method.

**Comment by DominikGuzei** [comment]
`define` static method? you mean like `Space.messaging.define`?

**Comment by rhyslbw** [comment]
@DominikGuzei Yep, but as you pointed out in Slack, we could support object field props, which is better

**Comment by DominikGuzei** [comment]
Yep, that's definitely an option. This would be an issue for people using ES6 / CS `class` though.

**Comment by Sanjo** [comment]
Yes, please make Aggregate.fields a method for consistency with the rest of space:base and space:domain.
